### ISSUE TITLE
Focal Loss returning NaN after some time of training with alpha=0.5 and gamma=0.5

### ISSUE THREAD

**Issue by fernandocamargoti** [fizyr/keras-retinanet#706]
After some time of the loss decreasing, it becomes NaN and the training stops converging.

**Comment by fernandocamargoti** [comment]
It seems that it was because of exploding gradients. I was setting clipnorm to be 0.001. After setting 0.0001, it stopped happening.

**Comment by YJHMITWEB** [comment]
Hi, I meet the same problem after this repo was updated. Do you get the same performance after training it with clipnorm==0.0001? e.g. mAP =35.0?

**Comment by Fellfalla** [comment]
This happens when the focal loss `gamma<1.0`, e.g. `0.5` or `0.0` and `classification = backend.gather_nd(classification, indices)` becomes `0.0` or `1.0` somewhere. For instance due to exploding gradients like in case of @fernandocamargoti.

So when the net output is 0.0 or 1.0 for any label, due to numerical errors, the focal_weight vector will contain negative values and `focal_weight ** gamma` results in `NaN` values

I propose to add `focal_weight = backend.clip(focal_weight, backend.epsilon(), 1.0)` before powering focal_weight:

```python
def _focal(y_true, y_pred):
    """ Compute the focal loss given the target tensor and the predicted tensor.
    As defined in https://arxiv.org/abs/1708.02002
    Args
        y_true: Tensor of target data from the generator with shape (B, N, num_classes).
        y_pred: Tensor of predicted data from the network with shape (B, N, num_classes).
    Returns
        The focal loss of y_pred w.r.t. y_true.
    """
    labels         = y_true[:, :, :-1]
    anchor_state   = y_true[:, :, -1]  # -1 for ignore, 0 for background, 1 for object
    classification = y_pred

    # filter out "ignore" anchors
    indices        = backend.where(keras.backend.not_equal(anchor_state, -1))
    labels         = backend.gather_nd(labels, indices)
    classification = backend.gather_nd(classification, indices)

    # compute the focal loss
    alpha_factor = keras.backend.ones_like(labels) * alpha
    alpha_factor = backend.where(keras.backend.equal(labels, 1), alpha_factor, 1 - alpha_factor)
    focal_weight = backend.where(keras.backend.equal(labels, 1), 1 - classification, classification)
    focal_weight = backend.clip(focal_weight, backend.epsilon(), 1.0) # avoid NaN when net output is 1.0 or 0.0
    focal_weight = alpha_factor * focal_weight ** gamma

    cls_loss = focal_weight * keras.backend.binary_crossentropy(labels, classification)

    # compute the normalizer: the number of positive anchors
    normalizer = backend.where(keras.backend.equal(anchor_state, 1))
    normalizer = keras.backend.cast(keras.backend.shape(normalizer)[0], keras.backend.floatx())
    normalizer = keras.backend.maximum(keras.backend.cast_to_floatx(1.0), normalizer)

    return keras.backend.sum(cls_loss) / normalizer

return _focal
```

**Comment by MartijnPeterVanLeeuwen** [comment]
I think the Nan comes from the eploding gradients when gamma<1. This comes from the fact that the derivative for the square root (originating from the gamma = 0.5) in the focal loss will approach infinity whenever x becomes extremely small. Try increasing the gamma to 1 or higher. It solved the issue for me.
